Magic V2 code doesn't compile under -DPERL_POISON

When building perl with  -DPERL_POISON, it now fails with
```
mg.c: In function ‘Perl_mg_free_struct’:
handy.h:3038:63: error: lvalue required as unary ‘&’ operand
 3038 |   ((d) ? (void)(safefree(MEM_LOG_FREE((Malloc_t)(d))), Poison(&(d), 1, Malloc_t)) : (void) 0)
```
This is because this line in mg.c:
```
            Safefree((char*)MgPTR(mg));
```
in addition to the actual freeing, also gets expanded to some code which fills the 8 bytes of the pointer itself with poison. It expands to something like:
```
              (void)memset(
                    (char*) (&(((void *)(mg)->mg_ptr))),
                    (U8)(0xEF),
                    (1) * sizeof(void *)
                )
```
and the (void*) cast in the MgPTR macro seems to upset '&'. Replacing the line in mg.c with a direct
```
Safefree(mg->mg_ptr);
```
fixes the build. But I don't know what the proper fix should be, as I don't know the intent of the (void*) cast in the MgPTR macro.

> When building perl with -DPERL_POISON, it now fails with
> 
> ```
> mg.c: In function ‘Perl_mg_free_struct’:
> handy.h:3038:63: error: lvalue required as unary ‘&’ operand
>  3038 |   ((d) ? (void)(safefree(MEM_LOG_FREE((Malloc_t)(d))), Poison(&(d), 1, Malloc_t)) : (void) 0)
> ```
> 
> This is because this line in mg.c:
> 
> ```
>             Safefree((char*)MgPTR(mg));
> ```
> 

I don't see that exact line in mg.c:
```
$ grep -in Safefree mg.c
961:            Safefree(MgPTR(mg));
977:                Safefree(mg->mg_ptr);
981:                Safefree(mg->mg_ptr);
988:    Safefree(mg);
1734:                Safefree(gary);
3071:        Safefree(free_me);
3329:        Safefree(mg->mg_ptr);
3347:        Safefree(mg->mg_ptr);
3362:    Safefree(mg->mg_ptr);	/* The mg_ptr holds the pos cache. */
3378:    Safefree(mg->mg_ptr);
3620:        Safefree(PL_inplace);
3629:            Safefree(PL_osname);
3735:            Safefree(IoTOP_NAME(io));
3746:            Safefree(IoFMT_NAME(io));
4005:            Safefree(gary);
```
Can you clarify?  thanks.

Confirmed, but slightly different error message:
```
In file included from perl.h:3381,
                 from mg.c:330:
mg.c: In function ‘Perl_mg_free_struct’:
handy.h:3038:63: error: lvalue required as unary ‘&’ operand
 3038 |   ((d) ? (void)(safefree(MEM_LOG_FREE((Malloc_t)(d))), Poison(&(d), 1, Malloc_t)) : (void) 0)
      |                                                               ^
handy.h:3080:76: note: in definition of macro ‘PoisonWith’
 3080 | #define PoisonWith(d,n,t,b)     (MEM_WRAP_CHECK_(n,t) (void)memset((char*)(d), (U8)(b), (n) * sizeof(t)))
      |                                                                            ^
handy.h:3083:33: note: in expansion of macro ‘PoisonFree’
 3083 | #define Poison(d,n,t)           PoisonFree(d,n,t)
      |                                 ^~~~~~~~~~
handy.h:3038:56: note: in expansion of macro ‘Poison’
 3038 |   ((d) ? (void)(safefree(MEM_LOG_FREE((Malloc_t)(d))), Poison(&(d), 1, Malloc_t)) : (void) 0)
      |                                                        ^~~~~~
mg.c:961:13: note: in expansion of macro ‘Safefree’
  961 |             Safefree(MgPTR(mg));
      |             ^~~~~~~~
make: *** [makefile:265: mg.o] Error 1
```

On Thu, Aug 20, 2026 at 07:02:37AM -0700, James E Keenan wrote:
> jkeenan left a comment (Perl/perl5#24736)
> 
> > When building perl with -DPERL_POISON, it now fails with
> > 
> > ```
> > mg.c: In function ‘Perl_mg_free_struct’:
> > handy.h:3038:63: error: lvalue required as unary ‘&’ operand
> >  3038 |   ((d) ? (void)(safefree(MEM_LOG_FREE((Malloc_t)(d))), Poison(&(d), 1, Malloc_t)) : (void) 0)
> > ```
> > 
> > This is because this line in mg.c:
> > 
> > ```
> >             Safefree((char*)MgPTR(mg));
> > ```
> > 
> 
> I don't see that exact line in mg.c:
> ```
> $ grep -in Safefree mg.c
> 961:            Safefree(MgPTR(mg));

That's because I accidentally cut+paste a temp edit. The real line is
that on line 961.

--- a/mg.c
+++ b/mg.c
@@ -958,7 +958,10 @@ Perl_mg_free_struct(pTHX_ SV *sv, MAGIC *mg)
             SvREFCNT_dec(MgKEYSV(mg));
 
         if(MgPTR(mg) && MgPTRLEN(mg)) {
-            Safefree(MgPTR(mg));
+            /* can't
+                Safefree(MgPTR(mg));
+              here because Safefree() requires an lvalue under -DPERL_POISON */
+            Safefree(mg->mg_ptr);
         }
 
         if(MgAUXSV(mg) && !MgWEAK_AUXSV(mg)) {

